qadence/backends/pyqtorch/convert_ops.py

```diff
@@ -222,165 +333,161 @@ def sparse_operation(state: Tensor, values: dict[str, Tensor] = None) -> Tensor:
         else:
 
-            def _fwd(state: Tensor, values: dict[str, Tensor]) -> Tensor:
+            def _hamiltonian(values: dict[str, Tensor]) -> Tensor:
                 hmat = _block_to_tensor_embedded(
                     block.generator,  # type: ignore[arg-type]
                     values=values,
-                    qubit_support=tuple(self.qubits),
+                    qubit_support=self.qubit_support,
                     use_full_support=False,
                 )
-                hmat = hmat.permute(1, 2, 0)
-                tevo = values[self.param_names[0]]
-                return self.operation(hmat, tevo, state)
+                return hmat.permute(1, 2, 0)
 
-        if config.use_gradient_checkpointing:
+            self._hamiltonian = _hamiltonian
 
-            def _forward(state: Tensor, values: dict[str, Tensor]) -> Tensor:
-                return checkpoint(_fwd, state, values, use_reentrant=False)
+        self._time_evolution = lambda values: values[self.param_names[0]]
 
-        else:
+    def _unitary(self, hamiltonian: Tensor, time_evolution: Tensor) -> Tensor:
+        self.batch_size = max(hamiltonian.size()[2], len(time_evolution))
+        diag_check = tensor([is_diag(hamiltonian[..., i]) for i in range(hamiltonian.size()[2])])
 
-            def _forward(state: Tensor, values: dict[str, Tensor]) -> Tensor:
-                return _fwd(state, values)
+        def _evolve_diag_operator(hamiltonian: Tensor, time_evolution: Tensor) -> Tensor:
+            evol_operator = diagonal(hamiltonian) * (-1j * time_evolution).view((-1, 1))
+            evol_operator = diag_embed(exp(evol_operator))
+            return transpose(evol_operator, 0, -1)
 
-        self._forward = _forward
+        def _evolve_matrixexp_operator(hamiltonian: Tensor, time_evolution: Tensor) -> Tensor:
+            evol_operator = transpose(hamiltonian, 0, -1) * (-1j * time_evolution).view((-1, 1, 1))
+            evol_operator = linalg.matrix_exp(evol_operator)
+            return transpose(evol_operator, 0, -1)
 
-    def forward(self, state: Tensor, values: dict[str, Tensor]) -> Tensor:
-        return self._forward(state, values)
+        evolve_operator = (
+            _evolve_diag_operator if bool(prod(diag_check)) else _evolve_matrixexp_operator
+        )
+        return evolve_operator(hamiltonian, time_evolution)
 
+    def unitary(self, values: dict[str, Tensor]) -> Tensor:
+        return self._unitary(self._hamiltonian(values), self._time_evolution(values))
 
-class AddPyQOperation(Module):
-    def __init__(
-        self, qubits: Sequence, n_qubits: int, operations: list[Module], config: Configuration
-    ):
-        super().__init__()
-        self.operations = operations
+    def jacobian_time(self, values: dict[str, Tensor]) -> Tensor:
+        return finitediff(
+            lambda t: self._unitary(time_evolution=t, hamiltonian=self._hamiltonian(values)),
+            values[self.param_names[0]],
+        )
```

Comment:
should we use `iH * e(-iHt)` here?

Reply:
yes for sure

Reply:
Yes, please.

Reply:
after discussing with @jpmoutinho , keeping this for now and handling it properly in a subsequent MR